### ISSUE TITLE
Problem with nested filter

### ISSUE THREAD

**Issue by tinivir** [philipn/django-rest-framework-filters#134]
Hi!

I downloaded last version of this library, django-filters, and django rest framework.

When I try to use nested serializing from manual I have next error:

KeyError: 'subtype__subtype'

```
from rest_framework import viewsets

import rest_framework_filters as filters

from rest_framework_filters.backends import DjangoFilterBackend

class SubtypeFilter(filters.FilterSet):
    class Meta:
        model = Subtype
        fields = {'subtype': ['exact', 'in', 'startswith']}

class TypeFilter(filters.FilterSet):
    subtype= filters.RelatedFilter(SubtypeFilter, name='subtype')

    class Meta:
        model = Type
        fields = {'type': ['exact', 'in', 'startswith']}

class TypeViewSet(viewsets.ModelViewSet):
    filter_backends = (DjangoFilterBackend,)
    filter_class = TypeFilter
    ...
```

url: ?subtype__subtype=value

**Comment by rpkilby** [comment]
Hi @tinivir - there isn't much we can do without more information. It would be helpful to see the models or the traceback generated from the API call. 

Off the cuff, the model doesn't make sense. Does your `Subtype` model have a `subtype` field?

**Comment by joshcrim** [comment]
I am having the same issue using the most recent versions of django-filters, rest-framework-filters, and rest-framework. Here's my traceback:

api/tests/test_filters.py:287: in test_eq
    % (self.project.name))
../local/lib/python2.7/site-packages/rest_framework/test.py:282: in get
    response = super(APIClient, self).get(path, data=data, **extra)
../local/lib/python2.7/site-packages/rest_framework/test.py:208: in get
    return self.generic('GET', path, **r)
../local/lib/python2.7/site-packages/django/test/client.py:379: in generic
    return self.request(**r)
../local/lib/python2.7/site-packages/rest_framework/test.py:279: in request
    return super(APIClient, self).request(**kwargs)
../local/lib/python2.7/site-packages/rest_framework/test.py:231: in request
    request = super(APIRequestFactory, self).request(**kwargs)
../local/lib/python2.7/site-packages/django/test/client.py:466: in request
    six.reraise(*exc_info)
../local/lib/python2.7/site-packages/django/core/handlers/base.py:132: in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
../local/lib/python2.7/site-packages/django/views/decorators/csrf.py:58: in wrapped_view
    return view_func(*args, **kwargs)
../local/lib/python2.7/site-packages/rest_framework/viewsets.py:83: in view
    return self.dispatch(request, *args, **kwargs)
../local/lib/python2.7/site-packages/rest_framework/views.py:477: in dispatch
    response = self.handle_exception(exc)
../local/lib/python2.7/site-packages/rest_framework/views.py:437: in handle_exception
    self.raise_uncaught_exception(exc)
../local/lib/python2.7/site-packages/rest_framework/views.py:474: in dispatch
    response = handler(request, *args, **kwargs)
../local/lib/python2.7/site-packages/rest_framework/mixins.py:40: in list
    queryset = self.filter_queryset(self.get_queryset())
../local/lib/python2.7/site-packages/rest_framework/generics.py:151: in filter_queryset
    queryset = backend().filter_queryset(self.request, queryset, self)
../local/lib/python2.7/site-packages/rest_framework_filters/backends.py:18: in filter_queryset
    return filter_class(request.query_params, queryset=queryset).qs
../local/lib/python2.7/site-packages/rest_framework_filters/filterset.py:325: in qs
    qs = super(FilterSet, self).qs
../local/lib/python2.7/site-packages/django_filters/filterset.py:335: in qs
    value = self.form.cleaned_data[name]
E   KeyError: 'project__name'

**Comment by rpkilby** [comment]
Hi @joshcrim - could you provide example models/filters/views?

**Comment by rpkilby** [comment]
I've also reran the test suite on travis, and no errors are reporting.

**Comment by joshcrim** [comment]
models.py
```python
class Project(models.Model):
    name = models.CharField("Project Name", max_length=30, db_index=True)

class Task(models.Model):
    project = models.ForeignKey(Project, blank=True, null=True, on_delete=models.SET_NULL)
```

filters.py
```python
class ProjectFilter(FilterSet):
    name = AllLookupsFilter(name='name')
    class Meta:
        model = Project
        fields = ('name',)

class TaskFilter(FilterSet):
    project = RelatedFilter(ProjectFilter, name='project')
    class Meta:
        model = Task
        fields = ( 'project',)
```

viewsets.py
```python
class ProjectViewSet(mixins.CreateModelMixin,
                                  mixins.RetrieveModelMixin,
                                  mixins.UpdateModelMixin,
                                  mixins.ListModelMixin,
                                  viewsets.GenericViewSet):
    model = Project
    serializer_class = ProjectSerializer
    filter_backends = (backends.DjangoFilterBackend,)
    filter_class = ProjectFilter
    queryset = Projects.objects.all()

class TaskViewSet(mixins.CreateModelMixin,
                               mixins.RetrieveModelMixin,
                               mixins.UpdateModelMixin,
                               mixins.ListModelMixin,
                               viewsets.GenericViewSet):
    model = Task
    serializer_class = TaskSerializer
    filter_backends = (backends.DjangoFilterBackend,)
    filter_class = TaskFilter
    queryset = Task.objects.all()
```

**Comment by joshcrim** [comment]
Dunno if the issues are related, but I was am getting unexpected results from some of the filters too. None of the filters were working at all when the settings file set the backend, but some work correctly now that I've placed `filter_backends` in the viewset, others still do not. 

`name__exact` returns all projects and doesnt filter anything
`name__iexact` filters as expected, case insensitive
`contains`, `startswith`, and `endswith` are all case insensitive but should be case sensitive

**Comment by rpkilby** [comment]
Thanks for the update. If you have the time, some sample data would be incredibly helpful. Otherwise, I'll try to work on a test case that demonstrates the issue. 

> None of the filters were working at all when the settings file set the backend, but some work correctly now that I've placed `filter_backends` in the viewset, others still do not.

Not sure what's going on there, but that would be an issue w/ DRF itself. Make sure that you've passed the full module path.

**Comment by joshcrim** [comment]
> Not sure what's going on there, but that would be an issue w/ DRF itself. Make sure that you've passed the full module path.

Gotcha, will do.

>  If you have the time, some sample data would be incredibly helpful.

Sample data?

**Comment by rpkilby** [comment]
> Sample data?

I mean example project names and the querystring you were using to filter. 

eg, "Project A", "Project B"

```http
GET https://localhost/api/projects?name=Project%20A
GET https://localhost/api/tasks?project__name=Project%20A
```

**Comment by rpkilby** [comment]
Oh - just noticed something: You're using `name__exact` when you should be filtering by just `name`.

**Comment by joshcrim** [comment]
Right, not using `name__exact` anymore for that reason ^_^;

**Comment by joshcrim** [comment]
As for data...the production query that set off an alert was this:

```
GET http://prodsite//task?description__iexact=200%20-%20Project%20Management&project__name__iexact=00119163383
```

I wrote my own test cases to reproduce the issue and narrowed it down to `?project__name` being where it breaks. `?project__iexact`, `?description__iexact`, and `?description__iexact='x'&project__iexact='x'` all work fine (using the projects id/tasks description in the query, of course). Throwing project__name in there is the only thing that breaks it.

```
GET https://devsite/task?project__name='a project name'
```

**Comment by joshcrim** [comment]
Here is another traceback from sentry.io that may be helpful. It is much longer and has a bit more detail:

http://pastebin.com/7ETvETET

**Comment by rpkilby** [comment]
Many thanks for all of the information. I will take a look at this later.

**Comment by joshcrim** [comment]
Quick update:
We downgrade our packages and everything is again working correctly:

```
django-rest-framework -- 3.5.2 --> 3.4.7
django-filters -- 0.15.3 --> 0.13
djangorestframework-filters -- 0.9.0 --> 0.8.0
```

**Comment by rpkilby** [comment]
Hm. I'd try downgrading packages individually and see what happens. Currently working on a test case

**Comment by rpkilby** [comment]
Hey @joshcrim - I've added a PR with the provided material. Can you review it? I'm still not able to recreate the issue.

**Comment by joshcrim** [comment]
Definitely, I'll try downgrading each package individually to see what breaks it and check the PR first thing in the morning.

**Comment by joshcrim** [comment]
We were initially on django-rest-framework 3.5.2 when we ran into this error, but have since downgraded back to 3.4.7. Getting the same results on both, so it doesnt look like drf version is an issue.

`django-rest-framework version 3.4.7`
```
df 0.13
rff 0.8
works correctly
```
```
df 0.14
rff 0.8
works correctly
```
```
df 0.15.3
rff 0.8
import error
```


```
df 0.13
rff 0.9
no __init__.py in django_filters/rest_framework
```
```
df 0.14
rff .9
import error
```
```
df 0.15.0
rff .9
project__name key error exception
```
```
df 0.15.3
rff 0.9
project__name key error exception
```

**Comment by joshcrim** [comment]
Ah, I forgot to send you my serializers. they are a little different. Everything else matches what I have

serializers.py
```python
class ProjectSerializer(serializers.HyperlinkedModelSerializer):
    name = serializers.CharField(max_length=30)

    class Meta:
        model = Project
        fields = ('name',)

class TaskSerializer(serializers.HyperlinkedModelSerializer):
    project = serializers.HyperlinkedRelatedField(
        queryset=Project.objects.all(),
        view_name='project-detail',
        required=False)

    class Meta:
        model = Task
        fields = ('project',)
```

**Comment by rpkilby** [comment]
Thanks for doing that. Those results make sense. django-filter 0.14 is compatible with drf-filters 0.8, and django-filter 0.15 is compatible with drf-filters 0.9

`0.14`/`0.8` is working, `0.15`/`0.9` is not.

**Comment by rpkilby** [comment]
A couple of ideas:
- Could you run your tests against django-filter 0.15.3 and the current master branch of this repo?
- Could you take a look at #137, see what differs?
- Try using the dict syntax with your filterset. eg,

```py
class ProjectFilter(FilterSet):
    class Meta:
        model = Project
        fields = {'name': '__all__'}
```

**Comment by joshcrim** [comment]
The only difference in #137  is the serializer fields: I am not using 'id' in either one. I do also use 'url' for both field tuples, but I dont think that makes a difference here.

Will try my tests against django-filter 0.15.3 and the current master branch later today. I believe I tried the dict syntax for fields already, but I'll gave it another go just to be sure.

**Comment by rpkilby** [comment]
Yeah, that's not going to make a difference here. The serializer won't affect the filtering process.

> Will try my tests against django-filter 0.15.3 and the current master branch later today. 

#127 was fixed on master, although it's a separate issue. That said, it's possible something has changed that fixed whatever you're running into. 

>  I believe I tried the dict syntax for fields already, but I'll gave it another go just to be sure.

There have been issues in the past where the dict syntax and the explicit `AllLookupsFilter` executed slightly different code paths.

**Comment by rpkilby** [comment]
Hi @joshcrim - did you have a chance to debug this any further? Thanks for the work you've put into this already.

**Comment by rpkilby** [comment]
Hi all, the issue seems to be an interaction with some crispy forms helper code in the DRF-compatible FilterSet. If you don't use crispy forms outside of DRF's browsable API, then the easiest short term fix would be to simply remove `'crispy_forms'` from your `INSTALLED_APPS`. If you do use crispy forms, then you could delete the offending constructor:

```python
from django_filters.rest_framework import FilterSet
del FilterSet.__init__
```

The longer term fix should be https://github.com/carltongibson/django-filter/pull/696.

**Comment by rpkilby** [comment]
Hi all, django-filter 1.1 has been released, as well as a compatible version of drf-filters.
